### PR TITLE
Deal with expired position the right way

### LINKED ISSUE #1071: Show loading screen when app starts with an expired position
So far we collaboratively close a position from the coordinator whenever a user comes online. 

This has the disadvantage that a user might not be aware of a protocol being executed and might close the app too fast which could break things. 

To overcome this we envision a loading screen when the user comes online which blocks the app for a short time and executes the protocol in the background.

### LINKED ISSUE #1173: Ensure that expired positions are updated on the maker
If a position is expired the coordinator should reduce the maker's position so that he can update his hedged position. 

Basically the coordinator needs to take an order of the opposite direction from the maker.

### PR DESCRIPTION
**Pull request by holzeis**
This PR got quite large and addresses several issues we had with closing expired positions.

1. Expired positions are now automatically matched against the orderbook

Since we currently have only one maker running which is publishing short and long orders this means the order for the expired position will get matched with the maker again, basically reducing his position with the coordinator.

2. Once the user comes online (connects to our websocket) we check for any pending matches (e.g. due to an expired position) and send a `AsyncMatch` message which the app is immediately executing.

For the following video, I reduced the position expiry to 30 seconds and the regular interval check for expired positions to 5 seconds. What happened in the background is that the orderbook found that the position expired found a match and saved that for later when the user is coming back online. Once online the match is executed with the user.

https://github.com/get10101/10101/assets/382048/ed0855bb-a732-4978-b24e-fc31ab8ab96e

resolves #1173 
resolves #1071

### CODE REVIEW

File: crates/orderbook-commons/src/lib.rs
Comment (bonomat) [inline crates/orderbook-commons/src/lib.rs:220]: Nit: 

That's a weird struct 🙈. Maybe make two out of them and pass them in separately? 
Alternatively, pass in the match and spare the creation of the new struct?

File: coordinator/migrations/2023-09-04-090152_add_order_state_and_matches/up.sql
Comment (bonomat) [inline coordinator/migrations/2023-09-04-090152_add_order_state_and_matches/up.sql:24]: Note: when running this the first time, all open not taken orders will be set to 'failed'. Not an issue because our maker will create new orders immediately, but I wanted to point it out :)

File: coordinator/src/node/expired_orders.rs
Comment (bonomat) [inline coordinator/src/node/expired_orders.rs:14]: Nit: 

this could be a single db call: 

```
UPDATE orders SET order_state = 'Failed' WHERE taken = false and  expiry < now();;
```

File: coordinator/src/orderbook/db/orders.rs
Comment (bonomat) [inline coordinator/src/orderbook/db/orders.rs]: Nit:

```suggestion 
let is_taken = order_state == orderbook_commons::OrderState::Taken;
```

File: coordinator/migrations/2023-09-04-090152_add_order_state_and_matches/up.sql
Comment (bonomat) [inline coordinator/migrations/2023-09-04-090152_add_order_state_and_matches/up.sql:29]: Changes in this file should be in an separate migration file or all in the same file.
If we end up deploying the commit which creates this file before this commit here, we will not re-run the migration and will get an error.

File: coordinator/migrations/2023-09-04-090152_add_order_state_and_matches/up.sql
Comment (bonomat) [inline coordinator/migrations/2023-09-04-090152_add_order_state_and_matches/up.sql:39]: Nit: 

could be a reference to user table

File: coordinator/src/node.rs
Comment (bonomat) [inline coordinator/src/node.rs:152]: Nit: 
```suggestion
            "Can't execute a trade on an expired order"
```

File: coordinator/src/node.rs
Comment (bonomat) [inline coordinator/src/node.rs:156]: Could be helpful to find future bugs if we print in what state the order was instead.

File: coordinator/src/node.rs
Comment (bonomat) [inline coordinator/src/node.rs]: You might want to handle the error of ` orders::set_order_state(` here instead of bubling it up or update both tables atomically, otherwise it may happen that one query is run and the other one isn't.

File: coordinator/src/node.rs
Comment (bonomat) [inline coordinator/src/node.rs:213]: Same comment as above. It's my understanding that these db updates should run atomically.

File: coordinator/src/node.rs
Comment (bonomat) [inline coordinator/src/node.rs:236]: Same again.

File: coordinator/src/node/expired_positions.rs
Comment (bonomat) [inline coordinator/src/node/expired_positions.rs:82]: is `%position.trader` printed nicely?

File: coordinator/src/node/expired_positions.rs
Comment (bonomat) [inline coordinator/src/node/expired_positions.rs:98]: 8 weeks is probably too long, we should force close sooner because of the CET's expiry but I understand where you want to get with this.

File: coordinator/src/orderbook/trading.rs
Comment (bonomat) [inline coordinator/src/orderbook/trading.rs:202]: image

File: coordinator/migrations/2023-09-04-090152_add_order_state_and_matches/up.sql
Comment (bonomat) [inline coordinator/migrations/2023-09-04-090152_add_order_state_and_matches/up.sql:17]: Same as in the other commit, please do not update old migration files. This can lead to annoying errors.

File: crates/orderbook-commons/src/order_matching_fee.rs
Comment (bonomat) [inline crates/orderbook-commons/src/order_matching_fee.rs:24]: MHmm... I don't like this. It should not happen that we get a match with `price=0`.

File: crates/orderbook-commons/src/order_matching_fee.rs
Comment (bonomat) [inline crates/orderbook-commons/src/order_matching_fee.rs:17]: Can you say when this can happen that the price is 0?

File: coordinator/src/orderbook/db/custom_types.rs
Comment (bonomat) [inline coordinator/src/orderbook/db/custom_types.rs:141]: Maybe we should only introduce these variants when we need/use them?

File: coordinator/src/orderbook/trading.rs
Comment (bonomat) [inline coordinator/src/orderbook/trading.rs:131]: Yes.

File: coordinator/src/orderbook/db/custom_types.rs
Comment (bonomat) [inline coordinator/src/orderbook/db/custom_types.rs:136]: `OrderReason` sounds weird to me. Is it more an `OrderType`?

File: coordinator/src/orderbook/trading.rs
Comment (bonomat) [inline coordinator/src/orderbook/trading.rs:284]: This might mess with us in the future: every non-manual order will be a `ForcedMatch`. Can you say this for sure? If not, maybe better to write all combinations out so that we see if a new variant was added.

File: coordinator/src/orderbook/db/custom_types.rs
Comment (bonomat) [inline coordinator/src/orderbook/db/custom_types.rs:137]: What's an `manual` order? 

Maybe add a description to what each variant means please?

File: mobile/lib/features/trade/force_order_change_notifier.dart
Comment (bonomat) [inline mobile/lib/features/trade/force_order_change_notifier.dart:62]: If this is the case, maybe add a log message if it happens?

File: mobile/native/src/db/models.rs
Comment (bonomat) [inline mobile/native/src/db/models.rs:226]: I'm not buying this concept of a _forced_ order yet... The naming does not align in my brain 🙈

**Review by bonomat — COMMENTED** [review]
Overall good work. 
It's not a good idea to change migration files, this makes commit not atomic and we could run into weird problems. 

I don't like the concept of a _forced_ order yet. You seem to define every order as _forced_ which was not _manual_. Maybe we can describe the concept differently? 

Otherwise, I think this looks good.

File: coordinator/src/orderbook/db/custom_types.rs
Comment (holzeis) [reply coordinator/src/orderbook/db/custom_types.rs:136]: We already have an `OrderType` 🙈

File: coordinator/migrations/2023-09-04-090152_add_order_state_and_matches/up.sql
Comment (holzeis) [reply coordinator/migrations/2023-09-04-090152_add_order_state_and_matches/up.sql:24]: I'll exclude limit orders

File: coordinator/src/node/expired_positions.rs
Comment (holzeis) [reply coordinator/src/node/expired_positions.rs:82]: > The % sigil operates similarly, but indicates that the value should be recorded using its [fmt::Display](https://doc.rust-lang.org/nightly/core/fmt/trait.Display.html) implementation:

https://docs.rs/tracing/latest/tracing/

File: coordinator/src/node/expired_positions.rs
Comment (holzeis) [reply coordinator/src/node/expired_positions.rs:98]: what would be a better value?

File: coordinator/src/node/expired_positions.rs
Comment (bonomat) [reply coordinator/src/node/expired_positions.rs:82]: That's not helpful 😅 will it be printed as `8b143` or as `[08,11,20,03,...` ? The latter one is not really helpful and we have heaps of them already, hence I'm asking.

File: coordinator/src/node/expired_positions.rs
Comment (bonomat) [reply coordinator/src/node/expired_positions.rs:98]: I'm not sure what our refund timeout is atm, but probably something like `refund_timeout-x` where x is a small number.

File: mobile/native/src/db/models.rs
Comment (holzeis) [reply mobile/native/src/db/models.rs:226]: renamed to async order

**Comment by holzeis** [comment]
> It's not a good idea to change migration files, this makes commit not atomic and we could run into weird problems. 

I will check how hard it is to adapt the commits to have a dedicated migration script per commit.

> I don't like the concept of a _forced_ order yet. You seem to define every order as _forced_ which was not _manual_. Maybe we can describe the concept differently? 

As discussed offline I renamed it to _async_ order.

File: crates/orderbook-commons/src/order_matching_fee.rs
Comment (holzeis) [reply crates/orderbook-commons/src/order_matching_fee.rs:24]: It definitely happened to me while developing, but I can drop that commit if you like.

File: crates/orderbook-commons/src/order_matching_fee.rs
Comment (holzeis) [reply crates/orderbook-commons/src/order_matching_fee.rs:17]: I think it happens when we only have a ask or bid price.

File: coordinator/src/orderbook/trading.rs
Comment (holzeis) [reply coordinator/src/orderbook/trading.rs]: I made it explicit, so we do not run into pitfalls when adding a new variant.

File: coordinator/src/node/expired_positions.rs
Comment (holzeis) [reply coordinator/src/node/expired_positions.rs:98]: @luckysori maybe you can help us out here? what is the refund timeout atm?

File: coordinator/migrations/2023-09-04-090152_add_order_state_and_matches/up.sql
Comment (holzeis) [reply coordinator/migrations/2023-09-04-090152_add_order_state_and_matches/up.sql]: I've split up the migration scripts into dedicated folders per commit.

File: coordinator/migrations/2023-09-04-090152_add_order_state_and_matches/up.sql
Comment (holzeis) [reply coordinator/migrations/2023-09-04-090152_add_order_state_and_matches/up.sql]: I've split up the migration scripts into dedicated folders per commit.

File: coordinator/src/orderbook/db/custom_types.rs
Comment (holzeis) [reply coordinator/src/orderbook/db/custom_types.rs:140]: How about SubOrderType?

File: crates/orderbook-commons/src/lib.rs
Comment (holzeis) [reply crates/orderbook-commons/src/lib.rs:216]: Unfortunately, that's not easily done. I need the relationship of the quantity to execution price. If I would split them up into two vecs I'd have to guess on the index that they relate, which is also not very nice.

Maybe improving the name of the struct helps? any suggestions?

File: coordinator/migrations/2023-09-04-090152_add_order_state_and_matches/up.sql
Comment (holzeis) [reply coordinator/migrations/2023-09-04-090152_add_order_state_and_matches/up.sql]: I though about it, but it we unfortunately do not register the maker as user, so we can't reference here the user table as it would otherwise fail to create the matches for the maker.

File: coordinator/src/orderbook/trading.rs
Comment (holzeis) [reply coordinator/src/orderbook/trading.rs:131]: I've created a ticket for that todo

https://github.com/get10101/10101/issues/1268

File: coordinator/migrations/2023-09-04-090152_add_order_state_and_matches/up.sql
Comment (holzeis) [reply coordinator/migrations/2023-09-04-090152_add_order_state_and_matches/up.sql]: I created a ticket for it

https://github.com/get10101/10101/issues/1269

**Review by bonomat — APPROVED** [review]
Looks good, but please wait for a second pair of eyes

**Review by luckysori — COMMENTED** [review]
Thanks for working on this long and complicated task!

Whilst I do like the concept of letting the coordinator post an order on behalf of the trader, who would most likely choose this over settling their position on-chain, the implementation is quite complicated and I am not convinced we are modelling things correctly with the introduction of `Matches` and `OrderReason`.

At this stage I don't really want to block the merge since other teammates have already approved this. My main worry is that the orderbook is becoming harder to understand and maintain, and I think it was already in a bad spot before this PR which adds more features.

File: coordinator/src/orderbook/trading.rs
Comment (holzeis) [reply coordinator/src/orderbook/trading.rs:161]: Market orders never have a price.

File: coordinator/src/node/expired_positions.rs
Comment (holzeis) [reply coordinator/src/node/expired_positions.rs:49]: How about adding the expiry to the matches table?

File: crates/orderbook-commons/src/lib.rs
Comment (holzeis) [reply crates/orderbook-commons/src/lib.rs:306]: That's the reason why I called it Matches! The issue is that those models should actually aligned. Unfortunately I couldn't map the model used as is to the database, that's the reason for the weird `matches` model.

File: coordinator/src/orderbook/trading.rs
Comment (holzeis) [reply coordinator/src/orderbook/trading.rs:225]: Because I want to track if a match got executed on per trader. The model would have gotten to complex otherwise.

Also considering that we could have multiple matches per order. By having a single match record we can easily reflect 1:n and m:n relations.

File: coordinator/migrations/2023-09-08-090152_add_order_state/up.sql
Comment (holzeis) [reply coordinator/migrations/2023-09-08-090152_add_order_state/up.sql:11]: already did in a later commit.

File: coordinator/src/node.rs
Comment (holzeis) [reply coordinator/src/node.rs:208]: I do agree that we should use transactions more consistently. I am opting to put that out of scope for that PR as the change is already quite big as is.

File: coordinator/src/node/expired_positions.rs
Comment (holzeis) [reply coordinator/src/node/expired_positions.rs:27]: Indeed its not needed :)

File: coordinator/src/bin/coordinator.rs
Comment (holzeis) [reply coordinator/src/bin/coordinator.rs:195]: don't you have to use it as it will otherwise get canceled, because it lost its scope?

File: coordinator/src/bin/coordinator.rs
Comment (holzeis) [reply coordinator/src/bin/coordinator.rs:194]: Maybe I am mistaken, but the `tx_price_feed` is received in the `websocket.rs` and is useful because with it I can broadcast the updated price to all connected users instead of having to send it to everyone individually.

```
// We subscribe *before* sending the "joined" message, so that we will also
// display it to our client.
let mut rx = state.tx_price_feed.subscribe();
```

cc: @bonomat

File: coordinator/src/orderbook/trading.rs
Comment (holzeis) [reply coordinator/src/orderbook/trading.rs]: Agreed, I am opting to not put that into scope of that PR.

File: coordinator/src/orderbook/trading.rs
Comment (holzeis) [reply coordinator/src/orderbook/trading.rs:180]: I think we still need it.. see comment on other discussion.

File: coordinator/src/orderbook/db/matches.rs
Comment (holzeis) [reply coordinator/src/orderbook/db/matches.rs:24]: I would have loved to name it `Match` unfortunately there is already a model named `Match`. We should have probably completely refactored the entire model, but instead I opted to simply go with a weird model name for now.

The refactor could come afterwards at any time I guess.

File: coordinator/migrations/2023-09-08-090152_add_order_state/up.sql
Comment (holzeis) [reply coordinator/migrations/2023-09-08-090152_add_order_state/up.sql:12]: Yes, that's the assumption! I think it's fair to do so as we are currently matching any market order immediately or set it to `Failed`. if there is a market order open which is taken false, means there must have been an error or the market order was submitted just about the time when we restarted and the order was not filled yet. In any case that would also mean that the order has `Failed` 🙃

File: crates/orderbook-commons/src/lib.rs
Comment (holzeis) [reply crates/orderbook-commons/src/lib.rs:162]: Unfortunately we have already a `Match` model and I did not want to refactor this model entirely. So I opted for the lesser effort, but definitely a bit uglier solution to name the model `Matches`. I think we can refactor that at a later point and then rename `Matches` to `Match`.

File: coordinator/src/bin/coordinator.rs
Comment (luckysori) [reply coordinator/src/bin/coordinator.rs]: That's the `RemoteHandle`.

**Pull request merged**
